lib/cisco_node_utils/environment.rb

```diff
@@ -0,0 +1,93 @@
+# March 2016, Glenn F. Matthews
+#
+# Copyright (c) 2016 Cisco and/or its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'yaml'
+require_relative 'logger'
+
+module Cisco
+  # Class representing the configuration environment
+  class Environment
+    @environments = {}
+    @default_environment_name = 'default'
+
+    class << self
+      attr_accessor :default_environment_name
+    end
+
+    # We have three tiers of configuration:
+    # 1) default (defined in this file)
+    # 2) System-wide gem configuration in /etc/cisco_node_utils.yaml
+    # 3) User configuration in ~/cisco_node_utils.yaml
+
+    DEFAULT_ENVIRONMENT = {
+      host:     nil, # localhost by default
+      port:     nil, # only applicable to gRPC
+      username: nil,
+      password: nil,
+    }
+
+    def self.environments
+      if @environments.empty?
+        @environments = merge_config('/etc/cisco_node_utils.yaml',
+                                     @environments)
+        @environments = merge_config('~/cisco_node_utils.yaml',
```

Comment:
Well, it overwrites any system values that are redefined by the per-user, yes - I think that's desirable. You can also use this to define *additional* per-user values (separate from but in addition to the system values) if you want.